### ISSUE TITLE
`Replacing docs for ...` on Windows

### ISSUE THREAD

**Issue by pfitzseb** [timholy/Revise.jl#161]
On Revise master:
```
julia> using Revise

julia> Revise.track(Base)

julia> 1+1
┌ Warning: Replacing docs for `Libdl.RTLD_DEEPBIND :: Union{}` in module `Libdl`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `Libdl.RTLD_FIRST :: Union{}` in module `Libdl`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `Libdl.RTLD_GLOBAL :: Union{}` in module `Libdl`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `Libdl.RTLD_LAZY :: Union{}` in module `Libdl`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `Libdl.RTLD_LOCAL :: Union{}` in module `Libdl`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `Libdl.RTLD_NODELETE :: Union{}` in module `Libdl`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `Libdl.RTLD_NOLOAD :: Union{}` in module `Libdl`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `Libdl.RTLD_NOW :: Union{}` in module `Libdl`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `LibGit2.Consts.GIT_MERGE :: Union{}` in module `LibGit2.Consts`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `LibGit2.Consts.GIT_MERGE_FILE_FAVOR :: Union{}` in module `LibGit2.Consts`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `LibGit2.Consts.GIT_MERGE_PREFERENCE :: Union{}` in module `LibGit2.Consts`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `LibGit2.Consts.GIT_MERGE_ANALYSIS :: Union{}` in module `LibGit2.Consts`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `LibGit2.Consts.GIT_REBASE_OPERATION :: Union{}` in module `LibGit2.Consts`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `LibGit2.Consts.GIT_REPOSITORY_OPEN :: Union{}` in module `LibGit2.Consts`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `LibGit2.Consts.GIT_OPT :: Union{}` in module `LibGit2.Consts`
└ @ Base.Docs docs\Docs.jl:223
┌ Warning: Replacing docs for `LibGit2.Consts.GIT_PROXY :: Union{}` in module `LibGit2.Consts`
└ @ Base.Docs docs\Docs.jl:223
2
```

**Comment by timholy** [comment]
This happens (unfortunately) every time you modify a method that has a docstring. I presume it's fixable but I haven't tried, it hasn't yet annoyed me sufficiently. This might be the impetus!

Do you have local modifications to any code in base? Or do I need to be worried that revisions are occurring that shouldn't be?

**Comment by pfitzseb** [comment]
No, this is on a fresh binary install without any local modifications. Seems to be windows specific since I don't see the same on Linux.

**Comment by timholy** [comment]
Hmm. OK, to facilitate my own debugging I've been meaning to set it up so that Revise can log its changes. Seems like that might be useful here. If you're amenable I'll push that change first and then you can use it and let me know what revisions are being made?

**Comment by pfitzseb** [comment]
Sure, just ping me and I'll help however I can.

**Comment by pfitzseb** [comment]
With just that patch I get
```
julia> Revise.track(Base)

julia> 1+1
ERROR: type Nothing has no field captures
Stacktrace:
 [1] getproperty(::Any, ::Symbol) at .\sysimg.jl:18
 [2] hash(::Revise.LineSkippingIterator, ::UInt64) at C:\Users\Basti\.julia\dev\Revise\src\relocatable_exprs.jl:150
```
since there are also symbols like `##358` or `#1643###x#418`. If I then just add a check to `continue` if the regex match is `nothing` I get [this](https://gist.github.com/pfitzseb/6757ea9dcfbc139ef9207416a1c94927).

With this slightly more sophisticated patch
```julia
function Base.hash(iter::LineSkippingIterator, h::UInt)
    h += hashlsi_seed
    for x in iter
        if x isa Symbol
            # For gensymmed symbols, hash on the "base" name
            xs = String(x)
            if startswith(xs, '#')
                cs = filter!(!isempty, split(xs, '#'))
                if isempty(cs) # this is never the case when tracking Base
                    @info xs
                    continue
                end
                h += hash(cs[end], h) # no idea if this is what you wanted to do with the regex
                continue
            end
        end
        h += hash(x, h)
    end
    h
end
```
I get [fewer debug messages](https://gist.github.com/pfitzseb/0f9dc30806f432c7e539f56e8c6b11dd).

There are still the `Replacing docs for ...` warnings around, so the patch didn't really work...

**Comment by timholy** [comment]
Thanks for pointing out that gensyms come in more flavors than I realized.

Actually, in both links, from the lines that are printed it seems to be working perfectly. All those log messages are of two types,
- the "Watching" methods indicate that the listed files have an `mtime` that is younger than when Julia was built. I guess if you download a binary that's inevitable. (I always build my Julia from source, and hence the source files are older than the build time.) That basically forces Revise to examine the content of the files. (How long does it take? For me `@time Revise.track(Base)` is ~0.25s. But I have only 3 files it examined.)
- the "Diff" lines all have empty expressions for `newexprs` and `oldexprs`. That means Revise found that there were no differences.

The fact that you're getting some `Replacing docs for ...` warnings indeed suggests that there are still issues. Something like
```julia
filter(r->r.message=="Diff" && (!isempty(r.kwargs[:newexprs]) || !isempty(r.kwargs[:oldexprs])), rlogger.logs)
```
might pull out the interesting entries.

**Comment by pfitzseb** [comment]
Ah yes... wasn't quite sure what I was looking for in the logs. [Here](https://gist.github.com/pfitzseb/4db991d1a4bb2b339a316ec4b58e4ac5)'s the output of that filter expression. Loads of gensyms in there again, but apparently only of the `##123` flavour.

With 
```julia
function Base.hash(iter::LineSkippingIterator, h::UInt)
    h += hashlsi_seed
    for x in iter
        if x isa Symbol
            # For gensymmed symbols, hash on the "base" name
            xs = String(x)
            if startswith(xs, '#')
                h += hash("gensym", h)
                continue
            end
        end
        h += hash(x, h)
    end
    h
end
```
I only get one log message that isn't filtered out:
```
julia> f[1]
Test.LogRecord(Debug, "Diff", Revise, "Parsing", :Revise_67b60259, "C:\\Users\\Basti\\.julia\\dev\\Revise\\src\\Revise.jl", 212, Base.Iterators.Pairs{Symbol,Any,Tuple{Symbol,Symbol,Symbol},NamedTuple{(:activemodule, :newexprs, :oldexprs),Tuple{Module,Set{Revise.RelocatableExpr},Set{Revise.RelocatableExpr}}}}(:activemodule=>Libdl,:newexprs=>Set(Revise.RelocatableExpr[Error showing value of type Test.LogRecord:
ERROR: deepcopy of Modules not supported
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] deepcopy_internal(::Module, ::IdDict{Any,Any}) at .\deepcopy.jl:34
...
```
Trying to get this to print at all with
``` 
function tryprint(r::Revise.RelocatableExpr)
    println(r.head)
    for arg in r.args
        if arg isa Revise.RelocatableExpr
            tryprint(arg)
        else
            try
                println(arg)
            catch e
                @error e
            end
        end
    end
    println()
end
```
yields [this](https://gist.github.com/pfitzseb/66a2ea26b0b6a6d5844712e58a267c18) (which is kinda hard to parse, but I didn't feel like spending my time on pretty printing). 
That's consistent with the fact that I'm getting the `Replacing Docs for...` warnings in `Libdl`. Let me know if you need any other info to debug this further.

---
Regarding timings: `Revise.track(Base)` takes about a quarter of a second for me as well, but a subsequent `1+1` takes something like 15s (can't time that with `@time` because the time seems to be spent in Revise's internals).

**Comment by timholy** [comment]
For anyone who stumbles across this issue, note the logging system has continued to evolve, so some of the commands above are outdated. Docs are at https://timholy.github.io/Revise.jl/latest/debugging.html. Also, the `deepcopy` is gone which should make things much easier.
